fili-core/src/main/java/com/yahoo/bard/webservice/web/filterparser/ApiFilterListener.java

```diff
@@ -0,0 +1,82 @@
+// Copyright 2017 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.web.filterparser;
+
+import java.util.Set;
+
+import com.yahoo.bard.webservice.data.dimension.Dimension;
+import com.yahoo.bard.webservice.data.dimension.DimensionDictionary;
+import com.yahoo.bard.webservice.data.dimension.DimensionField;
+import com.yahoo.bard.webservice.table.LogicalTable;
+import com.yahoo.bard.webservice.web.BadApiRequestException;
+import com.yahoo.bard.webservice.web.BadFilterException;
+import com.yahoo.bard.webservice.web.FilterOperation;
+import com.yahoo.bard.webservice.web.filterparser.FiltersParser.FilterContext;
+
+/**
+ * Listener to construct single ApiFilter objects.
+ */
+public class ApiFilterListener extends ApiFiltersListListener {
+    private Dimension dimension = null;
+    private DimensionField dimensionField = null;
+    private FilterOperation operation = null;
+    private Set<String> values = null;
+
+    /**
+     * Construct the listener.
+     *
+     * @param table LogicalTable to use for the filter
+     * @param dimensionDictionary Dictionary of dimensions
+     */
+    public ApiFilterListener(LogicalTable table, DimensionDictionary dimensionDictionary) {
+        super(table, dimensionDictionary);
+    }
+
+    /* (non-Javadoc)
+     * @see com.yahoo.bard.webservice.web.filterparser.ApiFiltersListListener#exitFilter(
+     * com.yahoo.bard.webservice.web.filterparser.FiltersParser.FilterContext)
+     */
+    @Override
+    public void exitFilter(FilterContext ctx) {
+        FiltersParser.FilterComponentContext f = ctx.filterComponent();
+        dimension = extractDimension(f);
+        if (dimension == null) {
```

Comment:
Was our old `ApiFilter` supposed to be able to support a `null` `dimension`? The code that tries to extract the dimension out of an `ApiFilter` certainly seems to assume that `dimension` will be not null. So I feel like we should either throw an exception here if `dimension` is `null` (so that we don't get a vague `NullPointerException` later), or the getters on these listeners (and even the `ApiFilter`?) should be returning `Optional`.